### PR TITLE
feature: add support for put requests

### PR DESCRIPTION
**Pull request by sbender9**
This allows plugins to register to handle PUT requests for specific paths and allows web service clients to receive put deltas for paths that they are updating.

### CODE REVIEW

File: lib/interfaces/ws.js [hunk 1]
```diff
@@ -98,6 +111,20 @@ module.exports = function (app) {
             return
           }
           app.handleMessage(spark.request.source || 'ws', msg)
+
+          msg.updates.forEach(update => {
+            update.values.forEach(valuePath => {
+              var currentSource = pathSources[valuePath.path]
+              if (_.isUndefined(currentSource) || currentSource != spark) {
```
Comment (tkurki) [inline lib/interfaces/ws.js]: If there is a previous spark for the path we should probably do something more. At least log something. Push a message down to the previous connection, maybe even disconnect it?

File: lib/put.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
+const _ = require('lodash')
+const debug = require('debug')('signalk-server:put')
+
+const pathPrefix = '/signalk'
+const versionPrefix = '/v1'
```
Comment (tkurki) [inline lib/put.js]: Don’t we have the version someplace already available?

Comment (tkurki) [reply lib/put.js]: Or the whole pathprefix

File: lib/put.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+const Result = {
+  success: 'SUCCESS',
+  failure: 'FAILURE'
+}
+
+const actionHandlers = {}
+const actions = {}
+var nextActionId = 1
```
Comment (tkurki) [inline lib/put.js:24]: How about uuid? That would work over restarts.

File: lib/put.js [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    app.get(apiPathPrefix + 'actions', function (req, res, next) {
+      res.json(actions)
+    })
+
+    app.get(apiPathPrefix + 'actions/:id', function (req, res, next) {
+      var action = actions[req.params.id]
+      if (!action) {
+        res.status(404).send()
+      } else {
+        res.json(action)
```
Comment (tkurki) [inline lib/put.js:45]: Everybody can see all actions?

File: lib/put.js [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+      res.json(info)
+    })
+  },
+
+  registerActionHandler: registerActionHandler,
+  actionResult: actionResult,
+  putPath: putPath
+}
+
+function putPath (app, fullPath, value) {
```
Comment (tkurki) [inline lib/put.js]: Acl?

File: lib/interfaces/ws.js (hunk 1 above)
Comment (sbender9) [reply lib/interfaces/ws.js]: Problem would be that you could have more than one valid connection updating the same path. For example, if you had two gps units.

File: lib/put.js (hunk 2 above)
Comment (sbender9) [reply lib/put.js]: I don't see it anywhere

File: lib/put.js (hunk 4 above)
Comment (sbender9) [reply lib/put.js:45]: Yes, currently anyone with read access can see all actions. Do you think we should do something else?

File: lib/interfaces/ws.js (hunk 1 above)
Comment (tkurki) [reply lib/interfaces/ws.js]: True. But pushing actions to the ws client that connected last seems too arbitrary to me. Store a poison value in pathsources, respond with a meaningful error in that case and figure out a way to additionalky specify the source?

File: lib/put.js (hunk 4 above)
Comment (tkurki) [reply lib/put.js:45]: I’m ok with that, especially with security by uuid obscurity.

Comment (tkurki) [reply lib/put.js:45]: I’m ok with that, especially with security by uuid obscurity.

**Comment by sbender9** [comment]
Sorry, being lazy with the commit.

I'll add the docs and rename.

Didn't think of binding some parameters for this. I'm remove `data`

File: lib/put.js (hunk 3 above)
Comment (tkurki) [reply lib/put.js:24]: Remove the dangling variable?

**Pull request merged**